streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import java.nio.ByteBuffer;
+import java.util.Base64;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Metadata to be committed together with TopicPartition offset
+ */
+public class TopicPartitionMetadata {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TopicPartitionMetadata.class);
+
+    // visible for testing
+    static final byte LATEST_MAGIC_BYTE = 2;
+
+    private final long partitionTime;
+    private final ProcessorMetadata processorMetadata;
+
+    static TopicPartitionMetadata with(final long partitionTime, final ProcessorMetadata processorMetadata) {
+        return new TopicPartitionMetadata(partitionTime, processorMetadata);
+    }
+
+    private TopicPartitionMetadata(final long partitionTime, final ProcessorMetadata processorMetadata) {
+        Objects.requireNonNull(processorMetadata);
+        this.partitionTime = partitionTime;
+        this.processorMetadata = processorMetadata;
+    }
+
+    public long partitionTime() {
+        return partitionTime;
+    }
+
+    public ProcessorMetadata processorMetadata() {
+        return processorMetadata;
+    }
+
+    public String encode() {
+        final byte[] serializedMeta = processorMetadata.serialize();
+        // Format: MAGIC_BYTE(1) + PartitionTime(8) + processMeta
+        final ByteBuffer buffer = ByteBuffer.allocate(Byte.BYTES + Long.BYTES + serializedMeta.length);
+        buffer.put(LATEST_MAGIC_BYTE);
+        buffer.putLong(partitionTime);
+        buffer.put(serializedMeta);
+        return Base64.getEncoder().encodeToString(buffer.array());
+    }
+
+    public static TopicPartitionMetadata decode(final String encryptedString) {
+        long timestamp = RecordQueue.UNKNOWN;
+        ProcessorMetadata metadata = ProcessorMetadata.emptyMetadata();
+
+        if (encryptedString == null || encryptedString.isEmpty()) {
```

Comment:
`encryptedString` cannot be `null` IIRC -- seems we should not check for this condition to no mask a potential bug, but rather fail-fast.